### PR TITLE
Fix Adagrad Optimizer (Add lr_scheduler and gradient_clipping)

### PR DESCRIPTION
**Pull request by sxjscience**
Fix Adagrad Optimizer. Add lr_scheduler and clip_gradient

### CODE REVIEW

**Comment by futurely** [comment]
The learning rate of of lr_scheduler isn't initialized correctly. SGD, SGLD, ccSGD and Adam all set it in the following way. The default learning rate is [0.1](https://github.com/dmlc/mxnet/blob/f74d74d325fcf879dffd6db78a8d54bfc9e06aa6/python/mxnet/lr_scheduler.py#L14).
```python
if lr_scheduler is not None:
            self.lr_scheduler.base_lr = learning_rate
```

**Comment by sxjscience** [comment]
Sorry for my carelessness. I'll make a PR for this.

**Comment by futurely** [comment]
RMSProp and AdaDelta have the same problem. The common logic should be part of the parent class.

**Comment by sxjscience** [comment]
I think it's reasonable to put `clip_gradient` into the parent class, but for the algorithms like AdaDelta, there is actually no idea of `learning_rate`.
